storage: Enable confidential ephemeral volumes

sprt opened the pull request:
###### Merge Checklist  <!-- REQUIRED -->
- [x] Followed patch format from upstream recommendation: https://github.com/kata-containers/community/blob/main/CONTRIBUTING.md#patch-format
  - [x] Included a single commit in a given PR - at least unless there are related commits and each makes sense as a change on its own.
- [x] Aware about the PR to be merged using "create a merge commit" rather than "squash and merge" (or similar)
- [x] genPolicy only: Ensured the tool still builds on Windows
- [x] The `upstream/missing` label (or `upstream/not-needed`) has been set on the PR. 

###### Summary <!-- REQUIRED -->
<!-- Quick explanation of WHAT changed and WHY. -->
Please see commit messages. Note that this approach follows the driver-shim interface design (with the new metadata driver options) that we agreed to propose upstream with Intel. When we sync our fork with upstream, the agent code will look much different, but those agent changes should be transparent to the end user.

Upstream proposal: https://github.com/confidential-containers/confidential-containers/issues/247

Depends on:
* Driver changes: #231 
* kata-packages-uvm-coco changes: https://github.com/microsoft/azurelinux/pull/9938
* kata-containers-cc update: TODO

Migration flows:
* Of course if you want confidentiality at all you'll need to upgrade to these PRs.
* If you only upgrade Kata but not your driver, the local volumes will still work, just not confidential.
* If you only upgrade your driver but not Kata, you're broken because you need the Kata bits and the driver is now hardcoded to be confidential.

###### Test Methodology
<!-- How was this test validated? i.e. local build, pipeline build etc. -->
Local testing

danmihai1 commented on src/agent/src/storage/mod.rs:
```diff
@@ -380,8 +381,122 @@ fn mount_storage(logger: &Logger, storage: &Storage) -> Result<()> {
     );
 
+    let confidential = storage.driver_options.contains(&"confidential=true".into());
+    let ephemeral = storage.driver_options.contains(&"ephemeral=true".into());
```
Would this work too?

contains("ephemeral=true".to_string());

That would be easier for me to read/understand. Is there some advantage of using into() here?

sprt replied:
Done

danmihai1 commented on src/agent/src/storage/mod.rs:
```diff
@@ -380,8 +381,122 @@ fn mount_storage(logger: &Logger, storage: &Storage) -> Result<()> {
+    let confidential = storage.driver_options.contains(&"confidential=true".into());
+    let ephemeral = storage.driver_options.contains(&"ephemeral=true".into());
+
+    let src_path = if confidential && ephemeral {
```
This if branch is too long. Would it be practical to move all/most of the if branch code into a separate function?

sprt replied:
Done

danmihai1 commented on src/agent/src/storage/mod.rs:
```diff
@@ -380,8 +381,122 @@ fn mount_storage(logger: &Logger, storage: &Storage) -> Result<()> {
+    let ephemeral = storage.driver_options.contains(&"ephemeral=true".into());
+
+    let src_path = if confidential && ephemeral {
+        // NOTE: This whole branch will be replaced by the following
```
I vote for a comment similar to:

// **TODO**: replace this branch with <...>/luks-encrypt-storage

danmihai1 commented on src/agent/src/storage/mod.rs:
```diff
@@ -380,8 +381,122 @@ fn mount_storage(logger: &Logger, storage: &Storage) -> Result<()> {
+        // Name of the devmapper that will live under /dev/mapper/.
+        let devmapper_device_name = src_path
+            .file_name()
+            .ok_or_else(|| anyhow!("could not get filename from {:?}", src_path))?
```
Does this easier to read version work too?

anyhow!("could not get filename from {src_path}")?

danmihai1 commented on src/agent/src/storage/mod.rs:
```diff
@@ -380,8 +381,122 @@ fn mount_storage(logger: &Logger, storage: &Storage) -> Result<()> {
+            .output()?;
+        if !output.status.success() {
+            info!(logger, "failed to encrypt device";
+                "status" => output.status.code().unwrap_or(-1),
```
Please confirm that you tested these error paths, and the log output looks reasonable. I don't remember seeing this info! syntax before...

sprt replied:
Confirmed. Copied this pattern from the same file.

sprt commented:
New changes:
* Addressed PR comments.
* Deploying the luks-encrypt-storage script from upstream at runtime: upstream checks the integrity tags in a lazy way (`--integrity-no-wipe`), and we need to copy that as the eager version has a significant performance overhead (~100sec for a 10GB per my testing). It was much easier to call a script rather than to convert all those commands to Rust `Command::new()` calls.

Next step: genpolicy changes.

sprt commented:
New changes: added genpolicy support.

Redent0r commented:
Looks good to me from what I understand. My only feedback (possibly for a different PR) would be how to go about testing this. Would it be something like:
1. Create pod with ephemeral volume
2. Write to it
3. Content of encrypted storage should not be available from the host side
?
